File: aoc2015/src/aoc2015/day02.clj
(ns aoc2015.day02)

(def data
  (map
   (fn [row] (map #(Integer. %) (clojure.string/split row #"x")))
   (clojure.string/split (slurp "resources/day02.input") #"\n")))

(defn paper [l w h]
  (let [a (* 2 l w)
        b (* 2 w h)
        c (* 2 h l)
        d (/ (min a b c) 2)]
    (+ a b c d)))

(defn ribbon [l w h]
  (let  [a (+ l l w w)
         b (+ w w h h)
         c (+ h h l l)] 
    (+ (* l w h) (min a b c))))

(defn answer-1 [] (reduce + 0 (map (fn [[a b c]] (paper a b c)) data)))

(defn answer-2 [] (reduce + 0 (map (fn [[a b c]] (ribbon a b c)) data)))

File: aoc2015/kotlin/src/main/kotlin/net/jondotcomdotorg/aoc/y2015/Day02.kt
package net.jondotcomdotorg.aoc.y2015

import util.AdventOfCode
import kotlin.math.min

object Day02 : AdventOfCode() {

    @JvmStatic
    fun main(args: Array<String>) {
        println("Day02:")
        println("\tPart 1: ${part1(inputFileLineSequence)}")
        println("\tPart 2: ${part2(inputFileLineSequence)}")
    }

    data class Package(val l: Int, val w: Int, val h: Int)

    private val Package.paper: Long
        get() = 2L * l * w + 2L * w * h + 2 * h * l + min(min(l * w, w * h), h * l)

    private val Package.ribbon: Long
        get() = listOf(l + w, w + h, h + l).minOrNull()!! * 2L + l * w * h

    private fun String.toPackage(): Package {
        return split("x")
            .let { (a, b, c) -> Package(a.toInt(), b.toInt(), c.toInt()) }
    }

    private fun part1(input: Sequence<String>) =
        input.sumOf { it.toPackage().paper }

    private fun part2(input: Sequence<String>) =
        input.sumOf { it.toPackage().ribbon }
}
